### ISSUE TITLE
Speeding up time to compile LFortran

### ISSUE THREAD

**Issue by certik** [lfortran/lfortran#5414]
Let's speedup the compilation of the C++ source code of LFortran. We identified at least the following potential speedups:

* [ ] split `asr.h` into smaller parts, and only use what we need; the stock `asr.h` should only contain the ASR definitions that are needed in all of the code base, but specialied visitors, possibly all visitors should be separate
* [ ] modify cmake to not build a library of libasr, but just collect all the files and build everything at once (currently the parallel compilation stalls waiting for the library to be created and linked --- single point bottleneck for some reason in cmake)
* [ ] get rid of some binaries that we have (cpptranslate, various parsing benchmarks), or not build them by default --- it seems link time is very slow.
* [ ] optimize our include header files, and try not to include 3rd party libs at all and minimize C++ stdlib includes (which can also be slow). Only include 3rd party stuff in .cpp files where the damage is limited.
* [ ] switch to ninja
* [ ] moving things from `.h` files to `.cpp` files. We have to be careful and benchmark the speed of compilation to ensure there is no regression --- some things have to be inlined. But I suspect most things can be moved to `.cpp` without a penalty.

Any other ideas? I bet we can speed up the compilation time by at least 50%.

Let's start with those, and then we'll see.

**Comment by pjh40** [comment]
I use Ninja all the time for builds... works fine.  My biggest wish is to eliminate `build0.sh` and just have it part of the CMake config, but maybe there are problems with Windows builds with that?

**Comment by certik** [comment]
Yes, we should figure out how to do this as part of cmake (it's actually already called as part of it, but there might be some corner cases when it fails). Note that we need some way to invoke it without running cmake when creating the release tarballs, so we probably want to keep the shell script.

On Windows I actually want to switch to https://github.com/prefix-dev/shell/, so that we can just maintain one bash script on all platforms.

**Comment by tanay-man** [comment]
Hi @pjh40 are you using ninja to build LFortran?

**Comment by pjh40** [comment]
Just a heads up: when running newer versions of Bison (e.g. 3.8.x), the `-r all` option found in `build0.sh` enables counterexample generation, which demonstrates shift/reduce and reduce/reduce conflicts.  While this is a pretty handy option for parser developers, it takes a _very_ long time to run.  These versions of Bison could use more specific options, like `-r itemset,lookahead,solved`, but I haven't tried that on older versions.

**Comment by certik** [comment]
We had a lot of issue with Bison. We can turn it off I think for default builds, maybe only run it at the CI or something.

**Comment by certik** [comment]
With Ninja I am getting on M4 Max:
```
$ time ninja lfortran  
[96/96] Linking CXX executable src/bin/lfortran
ninja lfortran  128.25s user 10.37s system 952% cpu 14.555 total
```
